email_scraper/src/main/java/EmailScraper.java

```diff
@@ -0,0 +1,139 @@
+            if (v.startsWith("https://")) {
+                continue;
+            }
+//            System.out.println(v);
+
+            try {
+                final Document doc = Jsoup.connect(v).get();
+                getEmails(doc);
+                ++link_count;
+
+                // visit links only if total number of visited links is not greater than NUMBER_OF_LINKS
+                if (link_count > NUMBER_OF_LINKS) {
+                    break;
+                }
+
+                final Elements links2go = doc.select("a[href]");
+                for (Element link : links2go) {
+                    final String link_str = link.attr("abs:href");
+
+                    // visit links only if not being visited
+                    if (links.add(link_str)) {
+//                        System.out.println(getDomainName(link_str));
+
+                        // visit links only if not sub-domain
+                        if (getDomainName(link_str).equals(domain)) {
+                            queue.add(link_str);
+//                            System.out.println(link_str);
+                        }
+                    }
+                }
+            } catch (IOException e) {
+                // handle "org.jsoup.HttpStatusException: HTTP error fetching URL. Status=404, ..."
+//                System.err.println("For '" + v + "': " + e.getMessage());
+            }
+        }
+    }
+
+    public String getDomainName(String url) {
+        try {
+            URI uri = new URI(url);
+            String domain = uri.getHost();
+            if (domain != null) {
+                return domain.startsWith("www.") ? domain : "www." + domain;
+            } else {
+                return "";
+            }
+        } catch (URISyntaxException e) {
+//            System.err.println("For '" + url + "': " + e.getMessage());
+            return "";
+        }
+    }
+
+    private void getEmails(Document doc) {
+        Pattern p = Pattern.compile("[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\\.[a-zA-Z0-9-.]+");
+        Matcher matcher = p.matcher(doc.text());
+        while (matcher.find()) {
+            String email_str = matcher.group();
+            if (email_str.endsWith(".")) {
+                email_str = email_str.substring(0, email_str.length() - 1);
+            }
+            emails.add(email_str);
+//            System.out.println(email_str);
+        }
+    }
+
+    private void printEmails() {
+        System.out.println("Found these email addresses:");
+        for (String email: emails) {
+            System.out.println(email);
+        }
+    }
+
+    public static void main(String[] args) {
+        // generate url starting with http://www.
+        // note: tentatively ignore https for simplicity
+        String url = new CliOpts(args).invoke();
+        if(!url.startsWith("www.")&& !url.startsWith("http://")){
+            url = "www."+url;
+        }
+        if(!url.startsWith("http://")){
+            url = "http://"+url;
+        }
```

Comment:
this code seems strange, it converts `jana.com` to `http://www.jana.com` but leaves `http://jana.com` alone. I'm not sure you need the `www` thing at all, just use the incoming domain.

Reply:
Good point. Unfortunately, jsoup seems to require `www` be present in URL so as to read document properly. Otherwise, jsoup would not be able to read document properly from URL.